pkg/app/piped/eventwatcher/eventwatcher.go

```diff
@@ -0,0 +1,251 @@
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+				continue
+			}
+			if err := w.updateValues(ctx, repo, cfg.Events, commitMsg); err != nil {
+				w.logger.Error("failed to update the values",
+					zap.String("repo-id", repoCfg.RepoID),
+					zap.Error(err),
+				)
+			}
+		}
+	}
+}
+
+// updateValues inspects all Event-definition and pushes the changes to git repo if there is.
+func (w *watcher) updateValues(ctx context.Context, repo git.Repo, events []config.EventWatcherEvent, commitMsg string) error {
+	commits := make([]*commit, 0)
+	for _, e := range events {
+		c, err := w.checkOutdatedValues(ctx, &e, repo, commitMsg)
+		if err != nil {
+			w.logger.Error("failed to check outdated value", zap.Error(err))
+			continue
+		}
+		if c != nil {
+			commits = append(commits, c)
+		}
+	}
+	if len(commits) == 0 {
+		return nil
+	}
+
+	w.logger.Info(fmt.Sprintf("there are %d outdated values", len(commits)))
+	// Copy the repo to another directory to avoid pull failure in the future.
+	tmpDir, err := ioutil.TempDir("", "event-watcher")
+	if err != nil {
+		return fmt.Errorf("failed to create a new temporary directory: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	tmpRepo, err := repo.Copy(filepath.Join(tmpDir, "tmp-repo"))
+	if err != nil {
+		return fmt.Errorf("failed to copy the repository to the temporary directory: %w", err)
+	}
+	for _, c := range commits {
+		if err := tmpRepo.CommitChanges(ctx, tmpRepo.GetClonedBranch(), c.message, false, c.changes); err != nil {
+			return fmt.Errorf("failed to perform git commit: %w", err)
+		}
+	}
+
+	return tmpRepo.Push(ctx, tmpRepo.GetClonedBranch())
+}
+
+// checkOutdatedValues gives back a change contents if any deviation exists between the value in
+// the given git repository and one in the control-plane.
+func (w *watcher) checkOutdatedValues(ctx context.Context, event *config.EventWatcherEvent, repo git.Repo, commitMsg string) (*commit, error) {
+	latestEvent, ok := w.eventGetter.GetLatest(ctx, event.Name, event.Labels)
+	if !ok {
+		return nil, fmt.Errorf("failed to get the latest Event with the name %q", event.Name)
+	}
+
+	// Determine files to be changed.
+	changes := make(map[string][]byte, 0)
+	for _, r := range event.Replacements {
+		path := filepath.Join(repo.GetPath(), r.File)
+		yml, err := ioutil.ReadFile(path)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file: %w", err)
+		}
+		v, err := yamlprocessor.GetValue(yml, r.YAMLField)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get value at %s in %s: %w", r.YAMLField, r.File, err)
+		}
+		value, ok := v.(string)
+		if !ok {
+			return nil, fmt.Errorf("unknown type of value is defined at %s in %s", r.YAMLField, r.File)
+		}
+		if latestEvent.Data == value {
+			// Already up-to-date.
+			continue
+		}
+		// Put it into the change list.
+		newYml, err := yamlprocessor.ReplaceValue(yml, r.YAMLField, latestEvent.Data)
```

Comment:
In case the users configured multiple fields from the same file, will this override the old changes?

Reply:
Yes, that commit will happen conflict and end with failure.

Reply:
But I assumed it doesn't happen if the latest values are the same. It happens when only the latest values are different from each other.

Reply:
Ah, I just got what you mean. No. it always override the original content, defined in Git.

Reply:
All of my decisions are based on the rule that never makes a change that crosses lines, but you know what I mean?

Reply:
Ah, I totally got what you're most worried about. Yes, you're right. That can revert previous changes. Let me re-think it.

Reply:
I feel like we can simply solve it by editing a local file inside this loop.

Reply:
Yes. That is what I was talking about.

Reply:
haha, finally I caught up.

>I feel like we can simply solve it by editing a local file inside this loop.

I'm now inspecting some issues with this way.

Reply:
Looks like we have no problem with it, so applied it. PTAL 😄